# MQB7046 MODELLING PUBLIC HEALTH DATA
## ASSIGNMENT 1

1. Please complete your assignment and submit by 24 March 2024 (Sunday) 5pm.
2. Submit your notebook (.ipynb) in the UM spectrum. 
3. You can work either work independently or in a group of two. 


### Exercise 1
* Create two number variables to store a patients clinical measurements
    * Height was 1.82 m
    * Weight was 88.2 kg
* Calculate the BMI
* Use the round() function to round of the result to 2 decimals
    * If needed use help(round) to learn how to use this function
* Print the BMI and consider if the patient is obese.

In [1]:
# Create 2 number vairables
height = 1.82
weight = 88.2

# Calculate the BMI and round to 2 decimals
bmi = round((weight/(height**2)), 2)

# Define either patient is obese or not
# Values based on Malaysia CPG guideline management of obesity
# https://www.moh.gov.my/moh/resources/Penerbitan/CPG/Endocrine/CPG_Management_of_Obesity_(Second_Edition)_2023.pdf
if bmi >= 37.5:
    bmi_category = "Severely Obese"
elif bmi >= 32.5:
    bmi_category = "Moderately Obese"
elif bmi >= 27.5:
    bmi_category = "Obese"
elif bmi >= 23.0:
    bmi_category = "Overweight"
elif bmi >= 18.5:
    bmi_category = "Normal Weight"
else:
    bmi_category = "Underweight"

# Print out the the bmi category and patient's bmi
print("Patient is", bmi_category, "with BMI =", bmi)

Patient is Overweight with BMI = 26.63


### Exercise 2
* Using the build in string function to replace __z__ with __l__ and __b__ with __e__ in the string: _"you arb doing bxtrbmbzy wbzz"_
* Finally apply a function to capitalise the sentence

In [2]:
# Set a variable for the strings
sentence = "you arb doing bxtrbmbzy wbzz"

# Create a dictionary for string to be replace
word_replacer = {"z":"l", "b":"e"}

# Replace the string with looping on the dictionary above
for key, value in word_replacer.items():
    sentence = sentence.replace(key, value)

# Capitalize the sentence and print it out.
print(sentence.capitalize())

You are doing extremely well


### Exercise 3
* Your task is to create a Python program to convert Celsius temperatures to Fahrenheit. You can use a mixture of functions, for example, map() or lambda to achieve the desired result. You can perform the following steps.
  * Define a named function called _"celsius_to_fahrenheit"_ that takes a Celsius temperature as input and returns the equivalent temperature in Fahenheit.
  * Define a list named _"celsius"_ containing temperature values in Celcius.
  * Use the map() function along with the named function _"celsius_to_fahrenheit"_ to convert each Celsius temperature to Fahrenheit.
  * Store the converted temperatures in a new list named  _"fahrenheit"_.
  * 
Print both the original Celsius temperatures and the converted Fahrenheit temperatures.

In [3]:
# Define the function to convert Celsius to Fahrenheit
def celsius_to_fahrenheit(celsius):
    return celsius * 1.8 + 32

# Create a dummy list of Celsius
celsius = [0.0, 20.0, 40.0, 60.0, 80.0, 100.0]

# Convert the temperature list of celsius to fahrenheit
fahrenheit = list(map(celsius_to_fahrenheit, celsius))

# Print both list
print("Celsius list =", celsius)
print("Fahrenheit list =", fahrenheit)

Celsius list = [0.0, 20.0, 40.0, 60.0, 80.0, 100.0]
Fahrenheit list = [32.0, 68.0, 104.0, 140.0, 176.0, 212.0]


### Exercise 4
* Using the input function write a program that asks for a patient's weight in kg.
* Divide the weight with an appropriate factor to get the weight in pounds, and round to 1 decimal.
* Print out the string "The patient weight in pounds is:" followed by the weight in pounds.
* Use the functions int() and str() to make sure the data types are correct for what you want the code to do.

In [4]:
def convert_to_pounds(input_sentece: str, input_category: str):
        """To convert the input into float.
        If the user key in any non-numerical value will prompt an error to user and
        user are required to key in a new value again.

        Args:
            input_sentece (str): A sentence to ask what is the value user should key in for.
            input_category (str): A sentence to display a warning that user key in a non-numerical value

        Returns:
            float: input in float type
        """
        # Create an infinity loop for user to key in value
        while True:
            input_num = input(input_sentece.strip())
            try:
                # Try to convert the user input into float and return the value
                print(input_sentece, "User key in value =", input_num)
                return int(input_num)
                # A safety pin to break the infinite loop, despite after return python will stop the loop automatically
                break
            except:
                # Display a warning prompt if the value user key in is not numerical.
                print("Unable to convert", input_num, "into numerical value, please key in numerical numbers only for the", input_category)


# Ask for input of weight
weight_in_kg = convert_to_pounds("Please key in patient's weight in kg.", "patient's weight in kg.")
# Convert the weight from kg to pound, round it to 1 decimal & print it out
# Weight in pounds = weight in kg * 2.204622622 based on https://kilograms-to-pounds.com/
# Print the 
print("The patient weight in pounds is: {}".format(round(weight_in_kg*2.204622622, 1)))

Please key in patient's weight in kg. User key in value = 100kg
Unable to convert 100kg into numerical value, please key in numerical numbers only for the patient's weight in kg.
Please key in patient's weight in kg. User key in value = 100kg
Unable to convert 100kg into numerical value, please key in numerical numbers only for the patient's weight in kg.
Please key in patient's weight in kg. User key in value = 100
The patient weight in pounds is: 220.5


### Exercise 5

* [This page describes how to calculate the mean arterial pressure (MAP)](https://www.nursingcenter.com/ncblog/december-2011/calculating-the-map)
* Use this info to calculate the MAP of the patient whose blood pressure is as follows: SBP=120, DBP=80
* Print three statments:
    * One presenting the systolic and diastolic blood presure
    * One presenting the calculated MAP
    * And one stating _"MAP lower than recommended: "_ together with a boolean testing this using the threshold of the linked page. 

In [5]:
def map_calculator(sbp = 120, dbp = 80):
    """Calculate the mean arterial pressure with formula from 
    https://www.nursingcenter.com/ncblog/december-2011/calculating-the-map

    Args:
        sbp (int, optional): Please put your systolic blood pressure. Defaults to 120.
        dbp (int, optional): Please put your diastolic blood pressure. Defaults to 80.
    """
    # Calculate the mean arterial pressure with formula from
    # https://www.nursingcenter.com/ncblog/december-2011/calculating-the-map
    mean_arterial_pressure = (sbp+(2*dbp))/3

    # Print the systolic and diastolic blood pressure
    print("Systolic Blooc Pressure", sbp, "mmHg.")
    print("Diastolic Blooc Pressure", dbp, "mmHg.")

    # Deciding the MAP cattegory based on Surviving Sepsis Campaign, MAP should >= 65mmHg
    if mean_arterial_pressure >= 65:
        print("Your MAP is", str(mean_arterial_pressure) +"mmHg, which is within the recommendation. (Higher or equal than 65mmHg)")
    else:
        print("MAP lower than recommended:", mean_arterial_pressure)
    
# Perform the function
map_calculator(sbp=120, dbp=80)

Systolic Blooc Pressure 120 mmHg.
Diastolic Blooc Pressure 80 mmHg.
Your MAP is 93.33333333333333mmHg, which is within the recommendation. (Higher or equal than 65mmHg)


### Exercise 6
* Create a dictionary of lists for the following data. A patient can have 1 to 5 heart rate measurements:

| patid | HR1 | HR2 | HR3 | HR4 | HR5 |
| ----- | --- | ----| ----| ----| ----|
| 1001| 95 | 123 | 110 | 90 | 100 | 
| 1002 | 110 | 150 | 160 | 87 | 99 |
| 1003 | 120 | 120 | 110 | 200 | 150 | 140 |


* Print the full dictionary
* Print all the patients in the dictionary
* Print all measurements for patient 1003
* Delete all measurements for patient 1001
* Print the full dictionary after deletion.

In [6]:
# Create the dictionary for the data above
patients_data = {
    1001:[95, 123, 110, 90, 100],
    1002:[110, 150, 160, 87, 99],
    1003:[120, 120, 110, 200, 150]
}
print("Print all the patients in the dictionary")
print(patients_data)
print("Print all the measurements for patient 1003")
print(patients_data[1003])
print("Delete all measurements for patient 1001")
del(patients_data[1001])
print("Print the full dictionary after delet/ion")
print(patients_data)

Print all the patients in the dictionary
{1001: [95, 123, 110, 90, 100], 1002: [110, 150, 160, 87, 99], 1003: [120, 120, 110, 200, 150]}
Print all the measurements for patient 1003
[120, 120, 110, 200, 150]
Delete all measurements for patient 1001
Print the full dictionary after delet/ion
{1002: [110, 150, 160, 87, 99], 1003: [120, 120, 110, 200, 150]}


### Exercise 7

Write a Python program that prompts the user to enter 5 numbers one by one. Calculate and print the average of these 5 numbers.
You are required to:
1. Use a while loop to input numbers and calculate the total.
2. Ensure that the program handles invalid input gracefully, displaying an error message and prompting the user to enter a valid number.
3. Finally, calculate the average of the entered numbers and print it with two decimal places.

In [7]:
# Define the function for the above program
def average_of_5_numbers():
    """A simple function to allow users to key in 5 numerical values and 
    return the average of the 5 values user entered with 2 decimal places.
    """
    # Define the initial value with 0 first
    val = 0 # To calculate total values
    num = 1 # For loop with while

    # To loop the input for 5 times
    while num < 6:
        # To allow user to input
        input_num = input("Please input your value number " + str(num) +":")
        try:
            # To convert the input to float
            input_num = float(input_num.strip())
            print("User key in value of", input_num)
            # Add the input value into val
            val += input_num
            # Add 1 to num to prevent infinity loop
            num += 1
        except:
            # If the input_num unable to be converted to float
            print("The value you entered", input_num, "is not a valid number, please key in numbers only.")

    # Print the average values after round with 2 decimal places
    print("The average valuese for the 5 numbers you key in =", round(val/5, 2))

# Run the function
average_of_5_numbers()

The average valuese for the 5 numbers you key in = 9.4


### Exercise 8

You are tasked with creating a Python program to calculate the gross salary per week for employees in a factory. The factory policy dictates that employees receive 1.5 times their hourly rate for any hours worked beyond 40 hours per week. Your program should prompt the user to enter the number of hours worked and the hourly rate for an employee. The program should then calculate and display the gross salary per week based on the input provided.

Calculate the gross salary per week for Employee X if he works 55 hours with an hourly rate of 20.

In [8]:
# A function that ensure user input numeric value
def convert_to_float(input_sentece: str, input_category: str):
        """To convert the input into float.
        If the user key in any non-numerical value will prompt an error to user and
        user are required to key in a new value again.

        Args:
            input_sentece (str): A sentence to ask what is the value user should key in for.
            input_category (str): A sentence to display a warning that user key in a non-numerical value

        Returns:
            float: input in float type
        """
        # Create an infinity loop for user to key in value
        while True:
            input_num = input(input_sentece.strip())
            try:
                # Try to convert the user input into float and return the value
                print(input_sentece, "User key in value =", input_num)
                return float(input_num)
                # A safety pin to break the infinite loop, despite after return python will stop the loop automatically
                break
            except:
                # Display a warning prompt if the value user key in is not numerical.
                print("Unable to convert", input_num, "into numerical value, please key in numerical numbers only for the", input_category)

# Define a function to calculate the gross salary per week for employees
def gross_salary_per_week_calculator():
    """A simple program to calculate the gross salary per week for employee in the factory.
    The factory policy dictates that employees receive 1.5 times their hourly rate for any hours worked beyond 40 hours per week.
    User are required to enter the umber of hours worked and hourly rate for the employee user wish to calculatte
    This program will  only display the gross salary per week based on the input provided without return any value.
    """
    # Requesting user to input number of hours worked
    working_hours = convert_to_float("How many hours did you work in a week?", 
                                     "working hours per week.")
    
    # Requesting user to input hourly rate salary
    salary_per_hour = convert_to_float("What is your hourly rate salary?",
                                       "hourly rate salary.")

    # If worker worked beyong 40 yours per week, the salary should be 1.5 times their hourly rate beyond 40 hours
    if working_hours >= 40:
        # Calculate overtime salary
        overtime_salary = (working_hours - 40) * 1.5 * salary_per_hour
        # Add 40 hours salary with normal rate
        gross_salary_per_week = (40 * salary_per_hour) + overtime_salary
    else:
       # Calculate the gross salary per week without overtime
        gross_salary_per_week = salary_per_hour * working_hours
    
    # Display the gross salary for the employee
    print(f"""The gross salary per week for you is RM{gross_salary_per_week:,.2f}""")

# To input employee X who worked 55 hours with an hourly rate of 20
gross_salary_per_week_calculator()

How many hours did you work in a week? User key in value = 55
What is your hourly rate salary? User key in value = 20
The gross salary per week for you is RM1,250.00


### Exercise 9

* You are building a calorie tracker application. Write a Python program to calculate the total calories burned by a group of individuals during a workout session. The calorie burn is estimated based on the duration of the workout and the type of activity. Use lambda functions and conditionals to calculate the calorie burn for each individual.
  * Define a list of tuples where each tuple contains the name of the individual, the duration of the workout (in minutes), and the type of activity (e.g., running, cycling, swimming).
  * The following are the data from 5 individuals:
    * ("Ali", 30, "running"),
    * ("Ah kau", 45, "cycling"),
    * ("Samyob", 60, "swimming"),
    * ("Ahmad", 25, "swimming"),
    * ("Chin Hai", 40, "running")
  * Use any functions to calculate the calorie burn for each individual based on the following estimate:
    * Running: 700 calories per 60 minutes
    * Cycling: 650 calories per 60 minutes
    * Swimming: 575 calories per 60 minutes
  * Calculate and print the total calories burned by each individual.

In [9]:
# Define a list of typles where each tuple contains the name of the individual, the duration of the workout (in minutes), and the type of activity.
participants_list = [("Ali", 30, "running",),
                     ("Ah kau", 45, "cycling",),
                     ("Samyob", 60, "swimming",),
                     ("Ahmad", 25, "swimming",),
                     ("Chin Hai", 40, "running",)]

# Define a function to calculate the total calories burned out
def calculate_calories(activity = None, minutes = 0):
    # Defined a dictionary with the calories per activity per 60 minutes
    _calories_ratio = {"running":700,
                       "cycling":650,
                       "swimming":575}
    
    # To calculate the total calories, round the value to 2 decimal places.
    return _calories_ratio[activity]/60*minutes

# To loop through the list with the above function, round the value to 2 decimal places.
participants_list = [x + (calculate_calories(activity=x[2], minutes=x[1]),) for x in participants_list]

# Print all the calories intake for each participants via loop method
for participant in participants_list:
    print("Total calories burned out by", participant[0], "=", participant[3], "calories.")

Total calories burned out by Ali = 350.0 calories.
Total calories burned out by Ah kau = 487.5 calories.
Total calories burned out by Samyob = 575.0 calories.
Total calories burned out by Ahmad = 239.58333333333334 calories.
Total calories burned out by Chin Hai = 466.66666666666663 calories.


### Exercise 10

* You are tasked with creating a simple Python program to calculate the total calorie intake for a day based on the user's consumption of different food items. Create the program based on the following instructions:

1. Define a dictionary named food_items to store various food items as keys and their respective calorie values per serving as values. Use the following food items and their calorie values per serving:
    - "banana": 95
    - "yogurt": 100
    - "salad": 130
    - "chicken breast": 165
    - "rice": 112
    - "curry fish": 221
    - "stirfried kailan": 50
2. Prompt the user to input the number of servings consumed for each food item.
3. Calculate the total calorie intake for the day by multiplying the servings of each food item with its respective calorie value per serving and summing up the total calories. Print the total calorie intake for the day. Use the provided food items and their calorie values for your calculations.



In [10]:
# A function that ensure user input numeric value, repeated to prevent error in this cell that didn't run previous
def convert_to_float(input_sentece: str, input_category: str):
        """To convert the input into float.
        If the user key in any non-numerical value will prompt an error to user and
        user are required to key in a new value again.

        Args:
            input_sentece (str): A sentence to ask what is the value user should key in for.
            input_category (str): A sentence to display a warning that user key in a non-numerical value

        Returns:
            float: input in float type
        """
        # Create an infinity loop for user to key in value
        while True:
            input_num = input(input_sentece.strip())
            try:
                # Try to convert the user input into float and return the value
                print(input_sentece, "User key in value =", input_num)
                return float(input_num)
                # A safety pin to break the infinite loop, despite after return python will stop the loop automatically
                break
            except:
                # Display a warning prompt if the value user key in is not numerical.
                print("Unable to convert", input_num, "into numerical value, please key in numerical numbers only for", input_category)

# Define a dictionary named food_items to store various food items as keys and their respective calorie values per serving as values.
food_items = {
    "banana": 95,
    "yogurt": 100,
    "salad": 130,
    "chicken breast": 165,
    "rice": 112,
    "curry fish": 221,
    "stirfried kailan": 50,
}

# Initiate total calories as 0
total_calories_day = 0

# To loop through the list
for food, calories in food_items.items():
    # Prompt the user to input the number of servings consumed for each food item.
    number_of_serving = convert_to_float("How many " + food + " do you consumed?",
                                         "How many " + food + " do you consumed?")
    
    # Add the calories intake of the food into total_calories_day
    total_calories_day = total_calories_day + (number_of_serving * calories)

# Print the total calories intake
print("Total calories intake =", total_calories_day)

How many banana do you consumed? User key in value = 2
How many yogurt do you consumed? User key in value = 2
How many salad do you consumed? User key in value = 3
How many chicken breast do you consumed? User key in value = 1
How many rice do you consumed? User key in value = 3
How many curry fish do you consumed? User key in value = 1
How many stirfried kailan do you consumed? User key in value = 3
Total calories intake = 1652.0
